pkg/ruler/prom_client.go

```diff
@@ -0,0 +1,91 @@
+package ruler
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log/level"
+	promapi "github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+)
+
+const orgIDHeader = "X-Scope-OrgId"
+
+type promClient struct {
+	promAPI promv1.API
+	timeout time.Duration
+}
+
+func NewPromClient(queryFrontendUrl, prometheusHTTPPrefix, orgID string, timeout time.Duration) *promClient {
+	if queryFrontendUrl == "" {
+		return nil
+	}
+	client, err := promapi.NewClient(promapi.Config{
+		Address:      queryFrontendUrl + prometheusHTTPPrefix,
+		RoundTripper: &addOrgIDRoundTripper{orgID: orgID, next: http.DefaultTransport},
+	})
+	if err != nil {
+		// use distributor queryable instead
+		return nil
+	}
+
+	return &promClient{
+		promAPI: promv1.NewAPI(client),
+		timeout: timeout,
+	}
+}
+
+func (p *promClient) InstantQuery(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
+	log, ctx := spanlogger.New(ctx, "promClient.InstantQuery")
+	defer log.Span.Finish()
+
+	value, warns, err := p.promAPI.Query(ctx, qs, t, promv1.WithTimeout(p.timeout))
+	if err != nil {
+		level.Error(log).Log("err", err, "query", qs)
+		return nil, err
+	}
+	if len(warns) > 0 {
+		level.Warn(log).Log("warnings", strings.Join(warns, ", "), "query", qs)
+	}
+
+	switch value.Type() {
+	case model.ValVector:
+		vector := value.(model.Vector)
+		v := make([]promql.Sample, 0, len(vector))
+		for _, sample := range value.(model.Vector) {
+			metric := make([]labels.Label, 0, len(sample.Metric))
+			for k, v := range sample.Metric {
+				metric = append(metric, labels.Label{
+					Name:  string(k),
+					Value: string(v),
+				})
+			}
+			v = append(v, promql.Sample{
+				T:      sample.Timestamp.Unix(),
+				F:      float64(sample.Value),
+				Metric: metric,
+			})
+		}
+		return v, nil
+	default:
```

Comment:
Why it cannot be a scalar?

Reply:
I didn't know expr in rule could return scalar.. but I get to know like _scalar(count(my_metric_name == 0))_ return scalar. So, I will add it.